pkg/virtual/framework/dynamic/apiserver/discovery.go

```diff
@@ -330,3 +332,86 @@ func sortGroupDiscoveryByKubeAwareVersion(gd []metav1.GroupVersionForDiscovery)
 		return version.CompareKubeAwareVersionStrings(gd[i].Version, gd[j].Version) > 0
 	})
 }
+
+type openAPIHandler struct {
+	apiSetRetriever apidefinition.APIDefinitionSetGetter
+	delegate        http.Handler
+
+	openAPIV3Service *handler3.OpenAPIService
+}
+
+func newOpenAPIHandler(apiSetRetriever apidefinition.APIDefinitionSetGetter, delegate http.Handler) *openAPIHandler {
+	openAPIV3Service := handler3.NewOpenAPIService()
+
+	return &openAPIHandler{
+		apiSetRetriever: apiSetRetriever,
+		delegate:        delegate,
+
+		openAPIV3Service: openAPIV3Service,
+	}
+}
+
+func (h *openAPIHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// handle /openapi 								-> do nothing
+	// handle /openapi/v2 							-> do nothing for now. TODO: Implement.
+	// handle /openapi/v3							-> discovery endpoint
+	// handle /openapi/v3/apis/<group>/<version> 	-> serve OpenAPIV3Schema for <group>/<version>
+
+	pathParts := splitPath(r.URL.Path)
+
+	// if request doesn't start with /openapi return
+	// this check is a safety measure
+	if len(pathParts) == 1 && pathParts[0] != "openapi" {
+		h.delegate.ServeHTTP(w, r)
+		return
+	}
+
+	if len(pathParts) == 1 && pathParts[0] == "openapi" {
+		return
+	}
+
+	if len(pathParts) == 2 && pathParts[1] == "v2" { // handle /openapi/v2
+		// TODO: Implement OpenAPI V2 handler
+		return
+	} else if len(pathParts) == 2 && pathParts[1] == "v3" { // handle /openapi/v3
+		if err := h.fetchAndUpdate(r.Context(), ""); err != nil {
+			responsewriters.ErrorNegotiated(
+				apierrors.NewInternalError(fmt.Errorf("unable to  set: %w", err)),
+				errorCodecs, schema.GroupVersion{},
+				w, r)
+		}
+
+		h.openAPIV3Service.HandleDiscovery(w, r)
+		return
+	} else if len(pathParts) == 5 && pathParts[1] == "v3" && pathParts[2] == "apis" {
+		requestedGroup := pathParts[3]
+		if err := h.fetchAndUpdate(r.Context(), requestedGroup); err != nil {
+			responsewriters.ErrorNegotiated(
+				apierrors.NewInternalError(fmt.Errorf("unable to  set: %w", err)),
+				errorCodecs, schema.GroupVersion{},
+				w, r)
+		}
+
+		h.openAPIV3Service.HandleGroupVersion(w, r)
+	} else {
+		h.delegate.ServeHTTP(w, r)
+		return
+	}
+}
+
+func (h *openAPIHandler) fetchAndUpdate(ctx context.Context, group string) error {
+	apiDomainKey := dyncamiccontext.APIDomainKeyFrom(ctx)
+	apiSet, _, err := h.apiSetRetriever.GetAPIDefinitionSet(ctx, apiDomainKey)
+	if err != nil {
+		return err
+	}
+
+	for gvr, apiDefinition := range apiSet {
+		if group == "" || (group != "" && group == gvr.Group) {
+			spec := apiDefinition.GetOpenAPIV3Spec()
+			h.openAPIV3Service.UpdateGroupVersion(gvr.Group, spec)
```

Comment:
is this synchronous?

Reply:
It lazily updates the cache, IIUC.

Reply:
@sttts Do we prefer/require synchronous or asynchronous (lazy update) in this case?